[WPE] Allow to run layout-tests with the wayland display

clopez opened the pull request:
#### b198a2b9d6bc310d093dea65bfb105b16ab2d7a3
<pre>
[WPE] Allow to run layout-tests with the wayland display
<a href="https://bugs.webkit.org/show_bug.cgi?id=298719">https://bugs.webkit.org/show_bug.cgi?id=298719</a>

Reviewed by Nikolas Zimmermann.

This allows to see the browser window when running layout
tests with WPE if the optional --display-server=wayland
 parameter is passed.

Previously the code was always forcing the headless driver
(ignoring the user selection of display-server) and WTR
was not connecting to the wayland display.

This patch modifies the display-server drivers to pass the
corresponding WPE_DISPLAY environment variable that WPE uses
to select one or other display backend as default.

As a side effect of this, now is possible (for example) to run
WebDriver tests headless with without having to pass manually
the --headless flag to MiniBrowser.

* Tools/Scripts/webkitpy/port/headlessdriver.py:
(HeadlessDriver._setup_environ_for_test):
* Tools/Scripts/webkitpy/port/headlessdriver_unittest.py:
(HeadlessDriverTest.make_environment):
(HeadlessDriverTest.test_environment_needed_variables):
(HeadlessDriverTest.test_environment_forbidden_variables):
* Tools/Scripts/webkitpy/port/waylanddriver.py:
(WaylandDriver._setup_environ_for_test):
* Tools/Scripts/webkitpy/port/waylanddriver_unittest.py:
(WaylandDriverTest.make_environment):
(WaylandDriverTest.test_environment_needed_variables):
* Tools/Scripts/webkitpy/port/westondriver.py:
(WestonDriver._setup_environ_for_test):
* Tools/Scripts/webkitpy/port/westondriver_unittest.py:
(WestonDriverTest.test_driver_start):
(WestonDriverTest.test_driver_stop):
(WestonDriverTest):
(WestonDriverTest.test_environment_needed_variables):
(WestonDriverTest.test_start): Deleted.
(WestonDriverTest.test_stop): Deleted.
(WestonDriverTest.test_stop.FakeWestonProcess): Deleted.
(WestonDriverTest.test_stop.FakeWestonProcess.terminate): Deleted.
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort._driver_class):
(WPEPort.setup_environ_for_server):
* Tools/Scripts/webkitpy/port/wpe_unittest.py:
(WPEPortTest.test_get_browser_path):
(WPEPortTest):
(WPEPortTest.test_setup_environ_for_test_wpe_prefix):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
(WTR::PlatformWebViewClientWPE::PlatformWebViewClientWPE):
* Tools/glib/api_test_runner.py:
(TestRunner._setup_testing_environment_for_driver):
(TestRunner.run_tests):
(TestRunner._setup_testing_environment): Deleted.

Canonical link: <a href="https://commits.webkit.org/299949@main">https://commits.webkit.org/299949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f5bd7037b7cbdcdb60057ee90fb257eeb1708ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120620 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40314 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30966 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127012 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72695 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122496 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41011 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48891 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91610 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60876 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123572 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108121 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72159 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/119977 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26226 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70617 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102237 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129881 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47541 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36090 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100236 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47908 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100077 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45516 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23544 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44201 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19172 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47403 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46872 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50218 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48558 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -733,6 +733,9 @@ int main(int argc, char *argv[])
     }
 
+    // Enable headless mode also if the env var defined by the headless display-server driver is present
+    headlessMode = headlessMode || !g_strcmp0(getenv("WPE_USE_HEADLESS_VIEW_BACKEND"), "1");
```
We have WPE_DISPLAY that can be used to enforce a certain WPE platform backend, e.g. display-drm or headless. Doesn't that fit your use case?

clopez replied:
I have reworked the patch to rely on that env var

nikolaszimmermann commented on Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
```diff
@@ -42,8 +43,17 @@ IGNORE_CLANG_WARNINGS_END
 namespace WTR {
 
 PlatformWebViewClientWPE::PlatformWebViewClientWPE(WKPageConfigurationRef configuration)
-    : m_display(adoptGRef(wpe_display_headless_new()))
```
Ah we don't use the standard factory here so WPE_DISPLAY had no effect...

clopez replied:
The new patch changes that

nikolaszimmermann commented on Tools/Scripts/webkitpy/port/headlessdriver_unittest.py:
```diff
@@ -51,18 +51,22 @@ def make_driver(self, worker_number=0, xorg_running=False, executive=None):
+                            'WAYLAND_DISPLAY': 'wayland-0',
+                            'WAYLAND_SOCKET': 'wayland-0',
+                            'WPE_DISPLAY':  'wpe-display-drm',
+                            'WPE_DRM_DEVICE': 'drm1',
```
Why are you explicitly setting this? Furthermore, I think it needs to point to a proper absolute path.

clopez replied:
Ok, this is a unit test.
The idea here is to have a user environment as hostile as possible, so to ensure that the python display server driver does what is expected as possible even when the environment of the user is totally unexpected.

nikolaszimmermann replied:
Fair enough, it really doesn't matter for the unit test.

nikolaszimmermann commented on Tools/Scripts/webkitpy/port/wpe_unittest.py:
```diff
@@ -174,3 +174,24 @@ def test_get_browser_path(self):
+    def test_setup_environ_for_test_wpe_prefix(self):
+        environment_user = {'WPE_DISPLAY':  'wpe-display-drm',
+                            'WPE_DRM_DEVICE': 'drm1',
+                            'WPE_USE_EXPLICIT_SYNC': '1',
```
What is this variable about?

clopez replied:
Is just one of the variables available that i found by running `git grep "getenv.*WPE_" Source`. Just to make the unit test case use some env variables that are currently relevant

nikolaszimmermann commented on Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
```diff
@@ -42,9 +42,11 @@ IGNORE_CLANG_WARNINGS_END
-    m_view = WKViewCreate(m_display.get(), configuration);
+    WPEDisplay* display = wpe_display_get_default();
+    if (!display)
+        g_error("Failed to get the default WPE display\n");
```
Does it make sense to abort here?

carlosgcampos replied:
Yes, g_error() aborts

nikolaszimmermann replied:
Ah good to know.

nikolaszimmermann reviewed (approved):
LGTM!

clopez committed:
[WPE] Allow to run layout-tests with the wayland display
https://bugs.webkit.org/show_bug.cgi?id=298719

Reviewed by Nikolas Zimmermann.

This allows to see the browser window when running layout
tests with WPE if the optional --display-server=wayland
 parameter is passed.

Previously the code was always forcing the headless driver
(ignoring the user selection of display-server) and WTR
was not connecting to the wayland display.

This patch modifies the display-server drivers to pass the
corresponding WPE_DISPLAY environment variable that WPE uses
to select one or other display backend as default.

As a side effect of this, now is possible (for example) to run
WebDriver tests headless with without having to pass manually
the --headless flag to MiniBrowser.

* Tools/Scripts/webkitpy/port/headlessdriver.py:
(HeadlessDriver._setup_environ_for_test):
* Tools/Scripts/webkitpy/port/headlessdriver_unittest.py:
(HeadlessDriverTest.make_environment):
(HeadlessDriverTest.test_environment_needed_variables):
(HeadlessDriverTest.test_environment_forbidden_variables):
* Tools/Scripts/webkitpy/port/waylanddriver.py:
(WaylandDriver._setup_environ_for_test):
* Tools/Scripts/webkitpy/port/waylanddriver_unittest.py:
(WaylandDriverTest.make_environment):
(WaylandDriverTest.test_environment_needed_variables):
* Tools/Scripts/webkitpy/port/westondriver.py:
(WestonDriver._setup_environ_for_test):
* Tools/Scripts/webkitpy/port/westondriver_unittest.py:
(WestonDriverTest.test_driver_start):
(WestonDriverTest.test_driver_stop):
(WestonDriverTest):
(WestonDriverTest.test_environment_needed_variables):
(WestonDriverTest.test_start): Deleted.
(WestonDriverTest.test_stop): Deleted.
(WestonDriverTest.test_stop.FakeWestonProcess): Deleted.
(WestonDriverTest.test_stop.FakeWestonProcess.terminate): Deleted.
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort._driver_class):
(WPEPort.setup_environ_for_server):
* Tools/Scripts/webkitpy/port/wpe_unittest.py:
(WPEPortTest.test_get_browser_path):
(WPEPortTest):
(WPEPortTest.test_setup_environ_for_test_wpe_prefix):
* Tools/WebKitTestRunner/wpe/PlatformWebViewClientWPE.cpp:
(WTR::PlatformWebViewClientWPE::PlatformWebViewClientWPE):
* Tools/glib/api_test_runner.py:
(TestRunner._setup_testing_environment_for_driver):
(TestRunner.run_tests):
(TestRunner._setup_testing_environment): Deleted.

Canonical link: https://commits.webkit.org/299949@main
